App upgrade - addition of metric dashbaord to tomcat app. Fix for mem… (#4287)
SumoLogic/sumologic-documentation: docs/integrations/web-servers/opentelemetry/apache-tomcat-opentelemetry.md

<img src={useBaseUrl('img/integrations/web-servers/apache-tomcat.png')} alt="Thumbnail icon" width="65"/> <img src={useBaseUrl('img/send-data/otel-color.svg')} alt="Thumbnail icon" width="45"/>

The [Apache Tomcat](https://tomcat.apache.org/tomcat-8.5-doc/index.html) app is a unified logs and metrics app that helps you monitor the availability, performance, health, and resource utilization of your Apache Tomcat servers. Preconfigured dashboards provide insight into visitor locations, traffic patterns, errors, resource utilization, garbage collection, web server operations, and access from known malicious sources.

> Tomcat logs are sent to Sumo Logic through the Opentelemetry [filelog receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/filelogreceiver) and metrics are sent through the [JMX](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/jmxreceiver) receiver with the `target_system` set as [`tomcat`](https://github.com/open-telemetry/opentelemetry-java-contrib/blob/main/jmx-metrics/docs/target-systems/tomcat.md). 

Reviewer: ```suggestion
Tomcat logs are sent to Sumo Logic through the OpenTelemetry [filelog receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/filelogreceiver) and metrics are sent through the [JMX](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/jmxreceiver) receiver with the `target_system` set as [`tomcat`](https://github.com/open-telemetry/opentelemetry-java-contrib/blob/main/jmx-metrics/docs/target-systems/tomcat.md). 
```

<img src='https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/Tomcat-OpenTelemetry/Apache-Tomcat-Schematics.png' alt="Schematics" />

## Fields Created in Sumo Logic for Tomcat

The following are the [Fields](/docs/manage/fields) that will be created as part of the Tomcat App install, if not already present.

- `sumo.datasource`. Has a fixed value of `tomcat`.

> - `deployment.environment`. User configured. This is the deployment environment where the Memcache cluster resides. For example: `dev`, `prod` or `qa`.

Reviewer: ```suggestion
- `deployment.environment`. User configured. This is the deployment environment where the Tomcat cluster resides. For example: `dev`, `prod` or `qa`.
```

- `webengine.cluster.name`. User configured. Set a value of the cluster where your Tomcat instance resides. This will be tagged along with the data sent to Sumo Logic.
- `webengine.system`. Has a fixed value of `tomcat`.

> - `webengine.node.name`. This has the value of the FQDN of the machine where the OpenTelemetry collector is collecting logs and metrics from.

Reviewer: ```suggestion
- `webengine.node.name`. This parameter holds the value of the Fully Qualified Domain Name (FQDN) of the machine from which the OpenTelemetry collector is collecting logs and metrics.
```

### Prerequisites

#### For metric collection

JMX receiver collects Tomcat metrics from Tomcat server as part of the OpenTelemetry Collector (OTC).

>   1. Follow the instructions in [JMX - OpenTelemetry's prerequisites section](/docs/integrations/app-development/opentelemetry/jmx-opentelemetry/#prerequisites) to download the [JMX Metric Gatherer](https://github.com/open-telemetry/opentelemetry-java-contrib/blob/main/jmx-metrics/README.md) used by the [JMX Receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/jmxreceiver#details).

Reviewer: ```suggestion
  1. Follow the instructions in [JMX - OpenTelemetry's prerequisites section](/docs/integrations/app-development/opentelemetry/jmx-opentelemetry/#prerequisites) to download the [JMX Metric Gatherer](https://github.com/open-telemetry/opentelemetry-java-contrib/blob/main/jmx-metrics/README.md). This gatherer is used by the [JMX Receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/jmxreceiver#details).
```

### Step 2: Configure integration

In this step, you will configure the YAML required for Tomcat Collection.

The path of the log file configured to capture Tomcat logs is needed to be given here.

The files are typically located in `/usr/share/tomcat/logs/*`. If you're using a customized path, check the Tomcat.conf file for this information.

For metrics, you're required to provide the JMX endpoint (default is `localhost:11099`) along with the `collection_interval` (default is 1 minute).

> You can add any custom fields which you want to tag along with the data ingested in Sumo Logic. Click on the **Download YAML File** button to get the YAML file.

Reviewer: ```suggestion
You can add any custom fields which you want to tag along with the data ingested in Sumo Logic. Click the **Download YAML File** button to get the YAML file.
```

| transpose row _timeslice column component
```

## Sample metrics
```

> {"queryId":"A","_source":"apache-tomcat-otel","_metricId":"9kcuMkQW4hfMv_0DLpjnBg","webengine.node.name":"ip-54-163-207-30.ec2.internal","_sourceName":"Http Input","host.id":"i-071ade72d03650657","host.group":"tomcatHostGrp","telemetry.sdk.version":"1.38.0","os.type":"linux","sumo.datasource":"tomcat","_sourceCategory":"Labs/apache-tomcat-otel","deployment.environment":"dev","_contentType":"Carbon2","host.name":"ip-54-163-207-30.ec2.internal","metric":"tomcat.request_count","_collectorId":"000000000FAB6E04","_sourceId":"000000005A1D5A65","webengine.system":"tomcat","webengine.cluster.name":"tomcat_cluster","unit":"requests","telemetry.sdk.language":"java","telemetry.sdk.name":"opentelemetry","service.name":"unknown_service:java","_collector":"Labs - apache-tomcat-otel","proto_handler":"\"http-nio-8080\"","max":391,"min":144,"avg":213.86,"sum":2994,"latest":265,"count":14}

Reviewer: ```suggestion
{
  "queryId":"A",
  "_source":"apache-tomcat-otel",
  "_metricId":"9kcuMkQW4hfMv_0DLpjnBg",
  "webengine.node.name":"ip-54-163-207-30.ec2.internal",
  "_sourceName":"Http Input",
  "host.id":"i-071ade72d03650657",
  "host.group":"tomcatHostGrp",
  "telemetry.sdk.version":"1.38.0",
  "os.type":"linux",
  "sumo.datasource":"tomcat",
  "_sourceCategory":"Labs/apache-tomcat-otel",
  "deployment.environment":"dev",
  "_contentType":"Carbon2",
  "host.name":"ip-54-163-207-30.ec2.internal",
  "metric":"tomcat.request_count",
  "_collectorId":"000000000FAB6E04",
  "_sourceId":"000000005A1D5A65",
  "webengine.system":"tomcat",
  "webengine.cluster.name":"tomcat_cluster",
  "unit":"requests",
  "telemetry.sdk.language":"java",
  "telemetry.sdk.name":"opentelemetry",
  "service.name":"unknown_service:java",
  "_collector":"Labs - apache-tomcat-otel",
  "proto_handler":"\"http-nio-8080\"",
  "max":391,
  "min":144,
  "avg":213.86,
  "sum":2994,
  "latest":265,
  "count":14
}
```

```

## Sample metrics query

```
sumo.datasource=tomcat deployment.environment=* webengine.cluster.name=* webengine.node.name=* metric=tomcat.request_count  | sum by deployment.environment,webengine.cluster.name,webengine.node.name,proto_handler
```

## Viewing Tomcat dashboards

### Overview

> The **Apache Tomcat - Overview** dashboard provides a high-level view of information on visitor geographic locations, responses over time, and number of error codes and top urls causing error. In addition to this you can monitor  requests, error count and their rates, along with busy and idle threads for your Tomcat instance.

Reviewer: ```suggestion
The **Apache Tomcat - Overview** dashboard provides a high-level view of information on visitor geographic locations, responses over time, and number of error codes and top urls causing error. In addition to this, you can monitor requests, error count and their rates, along with busy and idle threads for your Tomcat instance.
```

> The **Apache Tomcat - Overview** dashboard provides a high-level view of information on visitor geographic locations, responses over time, and number of error codes and top urls causing error. In addition to this, you can monitor requests, error count and their rates, along with busy and idle threads for your Tomcat instance.

Reviewer: ```suggestion
The **Apache Tomcat - Overview** dashboard offers a high-level view of visitor geographic locations, responses over time, error codes, and the top URLs causing errors. Additionally, it allows you to monitor requests, error counts and rates, as well as busy and idle threads for your Tomcat instance.
```

#### Visitor Locations

> The **Apache Tomcat - Visitor Locations** dashboard provides a high-level view of Tomcat visitor geographic locations both worldwide and in the United States. Dashboard panels also show graphic trends for visits by country over time and visits by US region over time.

Reviewer: ```suggestion
The **Apache Tomcat - Visitor Locations** dashboard provides a high-level view of Tomcat visitor geographic locations both worldwide and in the United States. Dashboard panels also show graphic trends for visits by country over time and visits by U.S. region over time.
```

> The **Apache Tomcat - Visitor Locations** dashboard provides a high-level view of Tomcat visitor geographic locations both worldwide and in the United States. Dashboard panels also show graphic trends for visits by country over time and visits by U.S. region over time.

Reviewer: ```suggestion
The **Apache Tomcat - Visitor Locations** dashboard provides a high-level view of Tomcat visitor geographic locations both worldwide and in the U.S. Dashboard panels also show graphic trends for visits by country over time and visits by U.S. region over time.
```

### Web Server Operations

> The **Apache Tomcat - Web Server Operations** dashboard provides a high-level view combined with detailed information on the top ten bots, geographic locations and data for clients with high error rates, server errors over time, and non 200 response code status codes. Dashboard panels also show information on server error logs, error log levels, error responses by server, and the top URIs responsible for 404 responses.

Reviewer: ```suggestion
The **Apache Tomcat - Web Server Operations** dashboard provides a high-level view combined with detailed information on the top ten bots, geographic locations and data for clients with high error rates, server errors over time, and non 200-response code status codes. Dashboard panels also show information on server error logs, error log levels, error responses by server, and the top URIs responsible for 404 responses.
```

Use this dashboard to:

- To gain insights and understand threats in incoming traffic and discover potential IOCs. Incoming traffic requests are analyzed using the [Sumo Logic - CrowdStrike](/docs/integrations/security-threat-detection/threat-intel-quick-analysis/#threat-intel-faq) threat feed.

<img src='https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/Tomcat-OpenTelemetry/Apache-Tomcat-Threat-Intel.png' alt="Threat intel" />

### Connectors (metrics)

The **Apache Tomcat - Connectors** dashboard provides an at-a-glance view of error count, request count, request processing time, total bytes sent/received, total connections, and thread (bust and ideal) information.

> <img src='https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/Tomcat-OpenTelemetry/Apache-Tomcat-Connectors.png' alt="Threat intel" />

Reviewer 2: @sumoanema, this last image is broken when I view the PR locally. Could you verify that the URL is correct?

Author: Fixed
